internal/controller/reinstall/reinstall_test.go

```diff
@@ -0,0 +1,680 @@
+		Expect(err).NotTo(HaveOccurred())
+		Expect(clusterInstance.Status.Reinstall.InProgressGeneration).To(BeEmpty())
+	})
+
+	It("should set RequestEndTime if it is zero", func() {
+		clusterInstance.Status.Reinstall.RequestEndTime = metav1.Time{}
+		Expect(c.Update(ctx, clusterInstance)).To(Succeed())
+
+		err := handler.finalizeReinstallRequest(ctx, testLogger, clusterInstance)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(clusterInstance.Status.Reinstall.RequestEndTime.IsZero()).To(BeFalse())
+	})
+
+	It("should return an error if patching fails", func() {
+		errorMsg := "inject error"
+		c = fakeclient.NewClientBuilder().
+			WithInterceptorFuncs(interceptor.Funcs{
+				SubResourcePatch: func(ctx context.Context, client client.Client, subResourceName string, obj client.Object, patch client.Patch, opts ...client.SubResourcePatchOption) error {
+					return errors.New(errorMsg)
+				},
+			}).
+			WithScheme(scheme.Scheme).
+			WithStatusSubresource(&v1alpha1.ClusterInstance{}).
+			WithObjects(clusterInstance).
+			Build()
+
+		handler = &ReinstallHandler{
+			ConfigStore: configStore,
+			Client:      c,
+			Logger:      testLogger,
+		}
+
+		err := handler.finalizeReinstallRequest(ctx, testLogger, clusterInstance)
+		Expect(err).To(HaveOccurred())
+		Expect(err.Error()).To(ContainSubstring(errorMsg))
+	})
+})
+
+var _ = Describe("ensureValidReinstallRequest", func() {
+	var (
+		ctx        context.Context
+		c          client.Client
+		handler    *ReinstallHandler
+		testLogger *zap.Logger
+
+		clusterInstance *v1alpha1.ClusterInstance
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		testLogger = zap.NewNop().Named("Test")
+
+		configStore, err := configuration.NewConfigurationStore(configuration.NewDefaultConfiguration())
+		Expect(err).ToNot(HaveOccurred())
+
+		clusterInstance = &v1alpha1.ClusterInstance{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test",
+				Namespace: "test",
+			},
+			Spec: v1alpha1.ClusterInstanceSpec{
+				Reinstall: &v1alpha1.ReinstallSpec{
+					Generation: "1234",
+				},
+			},
+			Status: v1alpha1.ClusterInstanceStatus{
+				Conditions: []metav1.Condition{},
+				Reinstall:  &v1alpha1.ReinstallStatus{},
+			},
+		}
+
+		c = fakeclient.NewClientBuilder().
+			WithScheme(scheme.Scheme).
+			WithStatusSubresource(&v1alpha1.ClusterInstance{}).
+			WithObjects(clusterInstance).
+			Build()
+
+		handler = &ReinstallHandler{
+			ConfigStore: configStore,
+			Client:      c,
+			Logger:      testLogger,
+		}
+	})
+
+	It("should return an error if reinstalls are not allowed", func() {
+		handler.ConfigStore.SetAllowReinstalls(false)
+
+		_, err := handler.ensureValidReinstallRequest(ctx, testLogger, clusterInstance)
+		Expect(err).To(HaveOccurred())
+		Expect(err.Error()).To(ContainSubstring("siteConfig operator is not configured for cluster reinstalls"))
+	})
+
+	It("should set the request start time if not already set", func() {
```

Comment:
This is a bit of a strange side-effect for a validation method. Why are you doing this here?

Reply:
I agree with it being a "strange" side-effect... my intention was to only set the start time when the request was validated (and avoid having it updated). I'll move it out and into a separate task.